src/index/LocatedTriples.cpp

```diff
@@ -0,0 +1,243 @@
+      if (locatedTriple->shouldTripleExist_) {
+        // Insertion of a non-existent triple.
+        writeTripleToResult(*locatedTriple);
+      }
+      locatedTriple++;
+    } else if (cmpEq(locatedTriple, *row)) {
+      // locateTriple == row
+      if (!locatedTriple->shouldTripleExist_) {
+        // Deletion of an existing triple.
+        row++;
+      }
+      locatedTriple++;
+    } else {
+      // locateTriple > row
+      // The row is not deleted - copy it
+      *resultEntry++ = *row++;
+    }
+  }
+
+  if (locatedTriple != locatedTriples.end()) {
+    AD_CORRECTNESS_CHECK(row == block.end());
+    std::ranges::for_each(
+        std::ranges::subrange(locatedTriple, locatedTriples.end()) |
+            std::views::filter(&LocatedTriple::shouldTripleExist_),
+        writeTripleToResult);
+  }
+  if (row != block.end()) {
+    AD_CORRECTNESS_CHECK(locatedTriple == locatedTriples.end());
+    while (row != block.end()) {
+      *resultEntry++ = *row++;
+    }
+  }
+
+  result.resize(resultEntry - result.begin());
+  return result;
+}
+
+IdTable LocatedTriplesPerBlock::mergeTriples(size_t blockIndex,
+                                             const IdTable& block,
+                                             size_t numIndexColumns) const {
+  if (numIndexColumns == 3) {
+    return mergeTriplesImpl<3>(blockIndex, block);
+  } else if (numIndexColumns == 2) {
+    return mergeTriplesImpl<2>(blockIndex, block);
+  } else {
+    AD_CORRECTNESS_CHECK(numIndexColumns == 1);
+    return mergeTriplesImpl<1>(blockIndex, block);
+  }
+}
+
+// ____________________________________________________________________________
+std::vector<LocatedTriples::iterator> LocatedTriplesPerBlock::add(
+    const std::vector<LocatedTriple>& locatedTriples) {
+  std::vector<LocatedTriples::iterator> handles;
+  handles.reserve(locatedTriples.size());
+  for (auto triple : locatedTriples) {
+    LocatedTriples& locatedTriplesInBlock = map_[triple.blockIndex_];
+    auto [handle, wasInserted] = locatedTriplesInBlock.emplace(triple);
+    AD_CORRECTNESS_CHECK(wasInserted == true);
+    AD_CORRECTNESS_CHECK(handle != locatedTriplesInBlock.end());
+    ++numTriples_;
+    handles.emplace_back(handle);
+  }
+  return handles;
+};
+
+// ____________________________________________________________________________
+void LocatedTriplesPerBlock::erase(size_t blockIndex,
+                                   LocatedTriples::iterator iter) {
+  auto blockIter = map_.find(blockIndex);
+  AD_CONTRACT_CHECK(blockIter != map_.end(), "Block ", blockIndex,
+                    " is not contained.");
+  auto& block = blockIter->second;
+  block.erase(iter);
+  numTriples_--;
+  if (block.empty()) {
+    map_.erase(blockIndex);
+  }
+}
+
+void LocatedTriplesPerBlock::updateAugmentedMetadata(
+    const std::vector<CompressedBlockMetadata>& metadata) {
+  // Copy block metadata to augment it with updated triples.
+  IndexMetaDataMmap::BlocksType allBlocks{metadata};
```

Comment:
Why the complicated alias? Can't you just take the argument by value, mutate it in place, and then `std::move()`it into the `augmentedMetadata_`?